### ISSUE TITLE
The Pads goes into the ground

### ISSUE THREAD

**Issue by py7hon** [pioneerspacesim/pioneer#5154]
I cant docking because The pads goes into the ground on Canwa system, Canwa B, Anderson's Grave, Fort Alvarez.
[video](https://files.catbox.moe/ttra03.mkv)

__My pioneer version (and OS):__
ver 20210203 (142693b) on: Linux

**Comment by fluffyfreak** [comment]
Took a look at this but so far nothing I've tried has worked. There's a parameter to control the amount of allowed slope under the station but even tightening the tolerance for this doesn't seem to help.

Needs more investigation but on some planets I believe the only thing to do is actually flatten the terrain beneath the cities/stations.

**Comment by Gliese852** [comment]
@fluffyfreak do you think it would be difficult to implement that the terrain heights a read from the mesh of a certain model? Let's say in the station model, there be some mesh, not drawn, but indicating the terrain?

**Comment by fluffyfreak** [comment]
I think it needs something like https://github.com/fluffyfreak/pioneer/tree/RegionBasedTerrainControl which Ae-2222 wrote a loooong time ago, but with some performance tweaks and blending improvements.

**Comment by Web-eWorks** [comment]
@fluffyfreak I noticed the terrain heightfield is sampled continuously, and if you're sampling at a higher resolution than the terrain meshes you can get wildly varying results compared to the actual surface/slope of the terrain. This can lead to false positives when trying to calculate the slope of the terrain under a starport depending on how far distant the sample points are from the actual terrain mesh vertices.

**Comment by fluffyfreak** [comment]
Another location
![image](https://github.com/user-attachments/assets/86c98ded-8bc4-46f5-b660-4282713f2350)
Found by "Coronatrix HQ-0602 pilot" on Discord

**Comment by Bodasey** [comment]
Now / Version 9336ec4 affected: OPLI Contemplation in the Barnard's star system - a spaceport often used when starting new games...

Feature request: refuse assigned docking bay to get one not under ground.

**Comment by bszlrd** [comment]
Made a quick and dirty fix for this specific one: #6283 Changed the seed for the moon to avoid that bumpiness. Oddly enough the station was already relocated automatically to avoid it being underground, so slightly changing the coordinates didn't work.

**Comment by Bodasey** [comment]
multiple / inconsistent surface data (for Deimos) on runtime?

<img width="1280" height="748" alt="Image" src="https://github.com/user-attachments/assets/fe682d6c-10d6-46da-8d1b-d0a556dc80c4" />

<img width="1280" height="748" alt="Image" src="https://github.com/user-attachments/assets/2b20129e-f98c-4c78-ba35-4ddea5db301c" />

You may try to fulfill this rescue mission; the other ship needs 11t of hydrogen, but these ships to rescue are not broken apart or visibly damaged for usual, means they can be saved and are able to fly away when they have got fuel.

[2026-1-5-ship-under-ground.txt](https://github.com/user-attachments/files/25389732/2026-1-5-ship-under-ground.txt)

**Comment by fluffyfreak** [comment]
@Bodasey this might be because of the bug i introduced which is fixed in #6292

**Comment by Bodasey** [comment]
I don't think so: I have compiled your branch, have loaded the saved game - nothing has changed.

**Comment by Bodasey** [comment]
<img width="1370" height="797" alt="Image" src="https://github.com/user-attachments/assets/da5d23c2-0e65-486c-b669-e700872810a1" />

So when you can "tunnel" your ship through mountains without any damage, why not spaceports?

But these tunnels often only work near the tops of mountains, trying this at the mountains' base, you often get damages -> the mountaineous landscape (fractals?) are built in several steps, and the stations are set in too early? And also ship damage is calculated not on the last step?

**Comment by syonfox** [comment]
Hi guys was thinkin about this the last few days.  

I want to note that yeah its fun to tunnel your ship and figure out how to land on a covered pad.

I would relate this to another issue of when you are flying the collision detection is not tight on the ground. ie you can clip through small hills and all, this breaks immersion. when you see blocky stuff like above.

Also I was thinking that in the hawkens rock picture the how can we fix this?. the buildings are already rendering on the ground. 

So i guess there are 2 options move the landing pads like buildings up and down. but this might break autopilot.

<img width="1920" height="1080" alt="Image" src="https://github.com/user-attachments/assets/29a6863f-c511-43ac-95f8-bf2dbff9c97c" />

Above you can imagine a combination of lowering and raising the different pads and extending the beams to the ground. This would make it look 'real'.

But after drawing that I was thinking what do we do here on earth for construction sites. 

We do cut fill and make a flat surface for building. 

I think this might be the correct solution but it would involve editing the terrain mesh around the landing pad.


<img width="1920" height="1080" alt="Image" src="https://github.com/user-attachments/assets/537455ab-7cf3-4e2a-b5c7-57973c4e646f" />

I think doing this and creating a foundation pad in terrain would be the most realistic and proper solution

It would 

- solve pads being buried 

- solve pads floating randomly in space. 

- be realistic how people would construct space bases.

*Question* how can we locally modify the terrain mesh? where to start?


Ok thinking more and one last easy idea is to simply find the max height under the landing pad. 

raise the pad to that height. 

extend the pillars down to the ground if possible 

In very steep terrain this will cause the landing pad to be placed at the top of a mountain. 

But I think in the rendering engine a fill is easier then a cut. 


### Conclusion 

It would be nice if there was a way to increase collision detection precision so that the player cant clip through terrain. This would break the fun thing of hiding your ship in a rock on a base but also improve immersion as the ground will feel real. At the same time it would be nice to have a way of locally smoothing the terrain to a flat surface near the landing pads as if a construction team took a bulldozer and pushed the hills into the holes. 

But this is a lot of change to graphics rendering and all and Im not sure how feasible it is to fix without a major deep dive into the terrain system collision detection and city rendering. 

A simple fix is to just render a big gray concrete block and place the landing pad on top of it. above the highest local point.

What do you think. 

... just found pull #6086

I guess that was the original idea (local flattening) but it is maybe running into performance problems. Im on a ryzen laptop with 16gb ram maybe the simple fix may be faster for low end systems will poke around a little and test the https://github.com/fluffyfreak/pioneer/tree/regions-branch

**Comment by fluffyfreak** [comment]
I like the idea of placing each pad individually however that's not how the station models work. It's a single model/building that just looks like 6 pads for that particular model, the others are single buildings and they suffer from the same problem on steep terrain.

Yeah I need to sort that branch out, it was a reasonble start to the idea but I found problems with it, specifically it wouldn't always flatten the correct terrain! 😆

**Comment by syonfox** [comment]
Found related issue also. 

https://github.com/pioneerspacesim/pioneer/issues/4149

An another example close by at Robinson base.

<img width="1661" height="1040" alt="Image" src="https://github.com/user-attachments/assets/6d37791d-64af-43e5-9d32-b5c999a59c32" />

This above example is at least compiling a1fb325f4f7198d7c9b0bc12c4e83477b663fad2

https://github.com/syonfox/pioneer/compare/constalations...regions-branch

The above is helpful for looking at the commits / changes for this. I will have to do a lot of learning to wrap my head around it. 

It appears Geo Sphere extend Base Sphere. And Terrain Extend GeoSphere? Out of time for today but hope the merge helps later as its up to date. 

Does anyone know if there is documentation on the terain system it looks like a read the code but its dense math with few comments. 

```
class GeoSphere : public BaseSphere { // base sphere is before noise geo sphere 

It take SystemBody into the constructor. 

And Maintains the GeoPatches? I guess these are the actual tiles of for rendering.

```

Sorry for the scattered thought just need to note it down. Before I uncommented ApplySimpleHeightRegions
It completely changed the planet rendering making everything flatter maybe.

> specifically it wouldn't always flatten the correct terrain! 😆

as it flattens all of it but not enough. At Robinson the pad was still burred but the whole planet was less interesting.

My next step is to learn more haha. 

Some questions to answer.

- Where is the definitive source of truth for the terrain? (overriding GetHeights might cause confusion here)

- Are we correctly finding the regions? From the SystemBody we can find the surface starports.
At these locations we can set the terrain to the mean height.
with a interpolation 

The Region would then be starport footprint + interpolation distance

- When is the fundamental terrain generated at render time or at system load time?

**Comment by fluffyfreak** [comment]
I don't think that anyone has actually documented the terrain in a long time, and the last time was probably me 🤔 

- 'GetHeights' is the only source of truth for the terrain height, it goes off and does a full noise generation and is the only thing that accesses the terrain noise.
- I think there's bugs in the current Regions branch which means that it may not be finding the correct locations, this could be the maths is wrong (_but it works frequently too_), the data is wrong (_starports sometimes get relocated, perhaps it's the initial value instead of final location_), or something else
- Terrain is generated at runtime, asynchronously, using a multithreaded job system. An initial root sphere of 6 faces/GeoPatches is generated, then each face gets it own quadtree of GeoPatches. I suppose that root face is the root level of the GeoPatch quadtree is a better way of saying it.

I'll try to do a more detailed write-up ~on the dev-forum and link to it from here in a bit~

### BaseSphere/GeoSphere/GeoPatch/etc
The basic concept of the terrain is a Quadtrilateralized Spherical Cube, where the heightmaps are generated from noise at runtime.

`BaseSphere` : `GeoSphere` -> `GeoPatch` * 6
or
`BaseSphere` : `GasGiant` _// but we'll ignore this for now as it's not relevant here_

The `BaseSphere` owns the `Terrain` object which despite the name does not own any geometry or do any rendering, it's just the terrain noise generator. The `BaseSphere` itself is in turn owned by a `TerrainBody` which derives from `Body` and this is what lets it's exist in the system logic and recieve updates, get rendered, saved to Json etc etc.

The `GeoSphere` owns 6 `GeoPatch` objects in `m_patches` and does some, frankly, weird shit around making sure that there is data in the root `GeoPatch` which is totally my fault 😁 I blame the fact that this system evolved from a much older system and I kept some of that around conceptually!
In  `GeoSphere` update, once it's filled those root `GeoPatch`'s, it allows each `GeoPatch` to manage itself by calling their `LODUpdate` that will recurse down through the `GeoPatch` Quadtree. It's a Quadtree because each `GeoPatch` owns pointers to 4 child `GeoPatch` objects etc etc and they pretty much take care of themselves and any child objects.

The `GeoPatch` handles it's own and childrens LOD/existence, they request new `GeoPatch` data be generated from the `Terrain` object, and they recieve that data from the parent `GeoSphere` using the `GeoPatchID` to give every `GeoPatch` a unique ID.

The data generated for each `GeoPatch` is turned into geometry in the `UpdateVBOs` where a "skirt" of geometry is added to the mesh which consists of an outer ring of vertices which duplicate some of the outermost generated data like UVs, normals, colour, etc but the positions are below the generated height. This is done to avoid splits and gaps in the terrain caused by mismatched LODs being adjacent. 
I chose this method to replace the older method that kept LODs no more than 1 LOD level different between adjacent GeoPatches. That system imposed a lot of restrictions on the code, it had heavy performance penalties, and it required a lot of bookkeeping to know which patches were adjacent.
The new system with a skirt means that each `GeoPatch` is completely independent of it's neighbours and can be of totally different LOD levels, with the skirt hiding the gaps effectively enough.

**Comment by fluffyfreak** [comment]
Unfortunately the branch is out of date, and there's been some changes to how the terrain noise works so I'll rebase it asap

**Comment by fluffyfreak** [comment]
@syonfox added a short, hopefully useful 🤷, description of the GeoSphere/GeoPatch system above. It's probably not very good but I realise now that I've never tried to explain it to anyone else ever before 😆

**Comment by syonfox** [comment]
Thank you for the thoughtful response. 

> Terrain is generated at runtime, asynchronously, using a multithreaded job system. An initial root sphere of 6 faces/GeoPatches is generated, then each face gets it own quadtree of GeoPatches. I suppose that root face is the root level of the GeoPatch quadtree is a better way of saying it.

This is cool, it makes that the terrain is generated on the fly as you fly around a system because it would be infeasible to generate every pixel of a whole planet let alone system or galaxy...* on that note I really like how pioneer is seed based and maintains a consistent galaxy over time. 

\* Reminds me of the trade off between memory and compute 

When you enter a system the Body's of that system are instantiated. (is this true are they not innate to the system)

A `Body` has a `TerrainBody` has a `BaseSphere` 

```
(json data)
Body   ,
 \    ,        
SystemBody.......Station...CityOnPlanet  *edit (so a reason this is complicated is
    |                                            it is full chain from here
    V                                              all the way to here  (City to Terrain)
GeoSphere <---- Terrain ...(noise generator) . . . . . . . . . . .'
| | | ||\__(m_patches)                                               
| | | || \                                                           
6 Root GeoPatch----[Child GeoPatch 1,2,3,4]...(surfaces)



Note this is conceptual not 100% accurate read the code ;)  
```


Will need to update the above more but please let me know if it is wrong. 



### How to fix burred landing pads. 

## Strat one 

Move the landing pad up

1.1 How to move it 

## Strat 2 move the terrain down

2.2 how to move it

# Question what is best?

It depends on balance between realism and practicality. 

## Strat 3 

3.3 leave it as is 

- just a note why change a fun easter egg.  Having a pad be burred and you die is sad but also you check the ai next time. 


ok saving this before my computer restarts
